### PR TITLE
feat(CMF): mutate collections based on their ids

### PR DESCRIPTION
**Pull request by jsomsanith**
**Please check if the PR fulfills these requirements**

- [x] The commit(s) message(s) follows our
  [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md#commit-message-format)
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What is the current behavior?** (You can also link to an open issue here)
The CMF collection delete/mutation is based on the index in the collection.
This can be problematic : 
* the components have to keep the index of their model
* when collection change in component (ex: filter), the indexes are wrong.

**What is the new behavior?**
Collection deletion and mutation are reserved to List or Map collections.
In List, the items are objects with an `id` property
In Map, the key is the id

The 2 operations are based on the id.
The action waits for this operation format : 
```
const mutationAction = actions.collectionsActions.mutateCollection('datastores', {
    ...
    delete: ['old_element_id'],
    update: { 'existing_element_id': {id: 'existing_element_id', ... }, ... }
});
```

**Does this PR introduce a breaking change?**

- [x] Yes
- [ ] No

| The `mutateCollection` operations | Old format | New format |
|---|---|---|
| delete | Array of index | Array of item id |
| update | Map of key/value = index/item | Map of key/value = id/item |

### CODE REVIEW

File: packages/cmf/src/reducers/collectionsReducers.js [hunk 1]
```diff
@@ -54,6 +105,20 @@ function deleteCollectionElement(state, action) {
...
+			return updateMapElements(state, action);
+		}
+		else if (List.isList(collection)) {
+			return updateListElements(state, action);
+		}
+		else {
+			throw new Error('CMF collection update is only compatible with ImmutableJs List and Map');
+		}
+	}
+	return state;
```
Comment (jmfrancois) [inline packages/cmf/src/reducers/collectionsReducers.js]: the following code will not be executed

Comment (jsomsanith) [inline packages/cmf/src/reducers/collectionsReducers.js]: It is if `action.operations.delete` is falsy

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://264.talend.surge.sh)

File: packages/cmf/src/reducers/collectionsReducers.js (hunk 1 above)
Comment (acateland) [inline packages/cmf/src/reducers/collectionsReducers.js]: nope above return will prevent any of it since it at root scope of the function

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://264.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://264.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://264.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://264.talend.surge.sh)

File: packages/cmf/src/reducers/collectionsReducers.js (hunk 1 above)
Comment (jsomsanith) [inline packages/cmf/src/reducers/collectionsReducers.js]: sorry I thought you were talking about the return `above`.
Cleaned this dead code, thanks

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://264.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://264.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://264.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://264.talend.surge.sh)

**Pull request merged**